// @microsoft/fast-components-react-base/dist/card/card.js
import Foundation from "@microsoft/fast-components-foundation-react";
import { classNames } from "@microsoft/fast-web-utilities";
import React from "react";
import { DisplayNamePrefix } from "../utilities";
import { CardTag } from "./card.props";
class Card extends Foundation {
    constructor() {
        super(...arguments);
        this.handledProps = {
            children: void 0,
            managedClasses: void 0,
            tag: void 0,
        };
    }
    /**
     * Renders the component
     */
    render() {
        return (React.createElement(this.tag, Object.assign({}, this.unhandledProps(), { className: this.generateClassNames() }), this.props.children));
    }
    /**
     * Generates class names
     */
    generateClassNames() {
        return super.generateClassNames(classNames(this.props.managedClasses.card));
    }
    /**
     * Stores HTML tag for use in render
     */
    get tag() {
        return CardTag[this.props.tag] || CardTag.div;
    }
}
Card.displayName = `${DisplayNamePrefix}Card`;
Card.defaultProps = {
    managedClasses: {},
};
export default Card;
export * from "./card.props";
