// @microsoft/fast-components-react-base/dist/viewport-positioner/viewport-positioner.props.js
export var ViewportPositionerHorizontalPosition;
(function (ViewportPositionerHorizontalPosition) {
    ViewportPositionerHorizontalPosition["left"] = "left";
    ViewportPositionerHorizontalPosition["right"] = "right";
    ViewportPositionerHorizontalPosition["uncontrolled"] = "uncontrolled";
})(ViewportPositionerHorizontalPosition || (ViewportPositionerHorizontalPosition = {}));
export var ViewportPositionerVerticalPosition;
(function (ViewportPositionerVerticalPosition) {
    ViewportPositionerVerticalPosition["top"] = "top";
    ViewportPositionerVerticalPosition["bottom"] = "bottom";
    ViewportPositionerVerticalPosition["uncontrolled"] = "uncontrolled";
})(ViewportPositionerVerticalPosition || (ViewportPositionerVerticalPosition = {}));
export var AxisPositioningMode;
(function (AxisPositioningMode) {
    AxisPositioningMode["uncontrolled"] = "uncontrolled";
    AxisPositioningMode["adjacent"] = "adjacent";
    AxisPositioningMode["inset"] = "inset";
})(AxisPositioningMode || (AxisPositioningMode = {}));
