// @microsoft/fast-components-react-msft/dist/metatext/metatext.js
import Foundation from "@microsoft/fast-components-foundation-react";
import { TypographySize, TypographyTag } from "@microsoft/fast-components-react-base";
import { classNames } from "@microsoft/fast-web-utilities";
import React from "react";
import { Typography } from "../typography";
import { DisplayNamePrefix } from "../utilities";
import { MetatextTag, } from "./metatext.props";
class Metatext extends Foundation {
    constructor() {
        super(...arguments);
        this.handledProps = {
            managedClasses: void 0,
            tag: void 0,
        };
    }
    /**
     * Renders the component
     */
    render() {
        return (React.createElement(Typography, Object.assign({}, this.unhandledProps(), { tag: TypographyTag[this.props.tag], size: TypographySize._7, className: this.generateClassNames() }), this.props.children));
    }
    /**
     * Generates class names based on props
     */
    generateClassNames() {
        return super.generateClassNames(classNames(this.props.managedClasses.metatext));
    }
}
Metatext.defaultProps = {
    tag: MetatextTag.span,
    managedClasses: {},
};
Metatext.displayName = `${DisplayNamePrefix}Metatext`;
export default Metatext;
export * from "./metatext.props";
