// @microsoft/fast-jss-manager-react/dist/sheet-manager.js
import { jss, stylesheetRegistry } from "./jss";
/**
 * A class for managing instance of stylesheets. The SheetManager
 * tracks and associates compiled stylesheets with their style/design-system inputs,
 * as well as tracking number of times a style/design-system combination has been used.
 * Tracking this allows us to memoize compiled stylesheets and only compile a new sheet
 * when one does not already exist.
 */
export default class SheetManager {
    constructor() {
        this.jss = jss;
        this.registry = new WeakMap();
    }
    /**
     * Creates a new JSS stylesheet from a stylesheet and design-system.
     * If a JSS style sheet has been created with this stylesheet and design system already,
     * then simply track that another instance has been added
     */
    add(styles, designSystem, options) {
        const tracker = this.getTracker(styles, designSystem);
        if (Array.isArray(tracker)) {
            tracker[1]++;
            return;
        }
        let designSystemRegistry = this.registry.get(styles);
        if (designSystemRegistry === undefined) {
            // If we don't have any design system instances for the current stylesheet,
            // we need to create a new map to store them
            designSystemRegistry = new WeakMap();
            this.registry.set(styles, designSystemRegistry);
        }
        this.registry
            .get(styles)
            .set(designSystem, [this.createStyleSheet(styles, designSystem, options), 1]);
    }
    /**
     * Return the compiled jss stylesheet associated with a given style object and design system
     */
    get(styles, designSystem) {
        const tracker = this.getTracker(styles, designSystem);
        if (Array.isArray(tracker) && !!tracker[0]) {
            return tracker[0];
        }
        return;
    }
    /**
     * Removes a reference for a stylesheet and designSystem and adds a new reference. Useful
     * when the design system changes and the stylesheet should be associated with a new
     * design system
     */
    update(styles, previousDesignSystem, nextDesignSystem) {
        const tracker = this.getTracker(styles, previousDesignSystem);
        if (!tracker) {
            return;
        }
        /**
         * If we only have one sheet and this sheet isn't associated with the next design system, *and* it's an object,
         * we can simply update the sheet and move it's reference instead of completely tearing down the sheet and
         * re-creating a style element
         */
        if (tracker[1] === 1 &&
            !this.get(styles, nextDesignSystem) &&
            !!styles &&
            typeof styles === "object") {
            tracker[0].update(nextDesignSystem);
            this.registry.get(styles).delete(previousDesignSystem);
            this.registry.get(styles).set(nextDesignSystem, tracker);
        }
        else {
            this.remove(styles, previousDesignSystem);
            this.add(styles, nextDesignSystem, tracker[0].options);
        }
    }
    /**
     * Reduces the internal count for a stylesheet and designsystem. If the count becomes zero,
     * the sheet will be detached
     */
    remove(styles, designSystem) {
        const tracker = this.getTracker(styles, designSystem);
        if (Array.isArray(tracker)) {
            tracker[1]--;
            if (tracker[1] === 0) {
                const sheet = tracker[0];
                this.jss.removeStyleSheet(sheet);
                stylesheetRegistry.remove(sheet);
                this.registry.get(styles).delete(designSystem);
            }
        }
    }
    /**
     * Returns the number of components using a stylesheet with a designSystem
     */
    count(styles, designSystem) {
        const tracker = this.getTracker(styles, designSystem);
        if (Array.isArray(tracker)) {
            return tracker[1];
        }
        return -1;
    }
    /**
     * Removes all entries
     */
    clean() {
        this.registry = new WeakMap();
    }
    /**
     * Retrieve the sheet tracker tracking the styles and design system
     */
    getTracker(styles, designSystem) {
        const designSystemRegistry = this.registry.get(styles);
        if (designSystemRegistry instanceof WeakMap) {
            const tracker = designSystemRegistry.get(designSystem);
            if (Array.isArray(tracker)) {
                return tracker;
            }
        }
    }
    /**
     * Creates a JSS StyleSheet and attaches it to the DOM
     */
    createStyleSheet(styles, designSystem, options = {}) {
        const stylesheet = typeof styles === "function" ? styles(designSystem) : styles;
        const sheet = this.jss.createStyleSheet(stylesheet, Object.assign({ link: true }, options));
        sheet.update(designSystem).attach();
        stylesheetRegistry.add(sheet);
        return sheet;
    }
}
