// @microsoft/fast-components-styles-msft/dist/utilities/border.js
import { format, toPx } from "@microsoft/fast-jss-utilities";
import { cornerRadius, elevatedCornerRadius, focusOutlineWidth, } from "../utilities/design-system";
function cornerRadiusRule(value) {
    return { "border-radius": value };
}
/**
 * Sets the border radius for controls.
 */
export function applyCornerRadius() {
    return cornerRadiusRule(toPx(cornerRadius));
}
/**
 * Sets the border radius for elevated surfaces or controls.
 */
export function applyElevatedCornerRadius() {
    return applyFloatingCornerRadius();
}
/**
 * Sets the border radius for pill-shaped controls.
 */
export function applyPillCornerRadius() {
    return cornerRadiusRule("999px");
}
/**
 * @deprecated Use applyElevatedCornerRadius instead.
 */
export function applyFloatingCornerRadius() {
    return cornerRadiusRule(toPx(elevatedCornerRadius));
}
/**
 * Sets the border width, style, and color to reserve the space for the focus indicator.
 */
export function applyFocusPlaceholderBorder(config /* @deprecated - this function doesn't need an argument */) {
    return {
        border: format("{0} solid transparent", toPx(focusOutlineWidth)),
    };
}
