// @microsoft/fast-components-react-base/dist/text-field/text-field.js
import Foundation from "@microsoft/fast-components-foundation-react";
import { classNames } from "@microsoft/fast-web-utilities";
import React from "react";
import { DisplayNamePrefix } from "../utilities";
import { TextFieldType, } from "./text-field.props";
class TextField extends Foundation {
    constructor() {
        super(...arguments);
        this.handledProps = {
            managedClasses: void 0,
            type: void 0,
        };
    }
    /**
     * Renders the component
     */
    render() {
        return (React.createElement("input", Object.assign({}, this.unhandledProps(), { className: this.generateClassNames(), type: this.props.type || TextFieldType.text })));
    }
    /**
     * Generates class names
     */
    generateClassNames() {
        return super.generateClassNames(classNames(this.props.managedClasses.textField));
    }
}
TextField.displayName = `${DisplayNamePrefix}TextField`;
TextField.defaultProps = {
    managedClasses: {},
};
export default TextField;
export * from "./text-field.props";
