// @microsoft/fast-components-react-msft/dist/call-to-action/call-to-action.js
import Foundation from "@microsoft/fast-components-foundation-react";
import { callToActionButtonOverrides } from "@microsoft/fast-components-styles-msft";
import { classNames } from "@microsoft/fast-web-utilities";
import { get } from "lodash-es";
import React from "react";
import { Button, ButtonAppearance } from "../button";
import { DisplayNamePrefix } from "../utilities";
import { CallToActionAppearance, } from "./call-to-action.props";
class CallToAction extends Foundation {
    constructor() {
        super(...arguments);
        this.handledProps = {
            appearance: void 0,
            href: void 0,
            managedClasses: void 0,
            disabled: void 0,
        };
    }
    /**
     * Renders the component
     */
    render() {
        return (React.createElement(Button, Object.assign({}, this.unhandledProps(), { className: this.generateClassNames(), disabled: this.props.disabled, href: this.props.href, appearance: ButtonAppearance[CallToActionAppearance[this.props.appearance]], jssStyleSheet: callToActionButtonOverrides, afterContent: this.renderAfterContent() }), this.props.children));
    }
    /**
     * Generates class names
     */
    generateClassNames() {
        const { callToAction, callToAction__disabled, } = this.props.managedClasses;
        return super.generateClassNames(classNames(callToAction, [callToAction__disabled, this.props.disabled], [
            this.props.managedClasses[`callToAction__${this.props.appearance}`],
            !!this.props.appearance,
        ]));
    }
    renderAfterContent() {
        return () => {
            return (React.createElement("span", { className: get(this.props, "managedClasses.callToAction_glyph") },
                React.createElement("svg", { viewBox: "0 0 16 16", xmlns: "http://www.w3.org/2000/svg" },
                    React.createElement("path", { d: "M4.023 15.273L11.29 8 4.023.727l.704-.704L12.71 8l-7.984 7.977-.704-.704z" }))));
        };
    }
}
CallToAction.displayName = `${DisplayNamePrefix}CallToAction`;
CallToAction.defaultProps = {
    managedClasses: {},
};
export default CallToAction;
export * from "./call-to-action.props";
