// @microsoft/fast-animation/dist/fade/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var animateTo_1 = require("../animateTo");
/**
 * Key frame object for fade-in animations
 */
exports.fadeInKeyframes = [
    { opacity: "0.01" },
    { opacity: "1" },
];
/**
 * Key frame object for fade-out animations
 */
exports.fadeOutKeyframes = [
    { opacity: "1" },
    { opacity: "0" },
];
/**
 * EffectTiming defaults for fade animations
 */
exports.fadeEffectTiming = {
    easing: "linear",
    duration: 500,
};
/**
 * Creates an animation to fade an element into view
 */
function fadeIn(element, effectTiming) {
    if (effectTiming === void 0) { effectTiming = {}; }
    return applyFade(element, exports.fadeInKeyframes, effectTiming);
}
exports.fadeIn = fadeIn;
/**
 * Creates an animation to fade an element out of view
 */
function fadeOut(element, effectTiming) {
    if (effectTiming === void 0) { effectTiming = {}; }
    return applyFade(element, exports.fadeOutKeyframes, effectTiming);
}
exports.fadeOut = fadeOut;
function applyFade(element, keyframes, timing) {
    if (timing === void 0) { timing = {}; }
    var fadeAnimationTiming = Object.assign({}, exports.fadeEffectTiming, timing);
    var fadeAnimation = new animateTo_1.default(element, null, fadeAnimationTiming);
    fadeAnimation.addKeyframes(keyframes);
    return fadeAnimation;
}
exports.applyFade = applyFade;
