// @microsoft/fast-components-react-msft/dist/background/background.props.js
/**
 * Friendly names for the indexes of light mode backgrounds
 * on the neutral ramp. These values are designed to work with
 * the default neutral ramp provided by the DesignSystem
 *
 * @deprecated Use the recipes because they can be more dynamic for different ramps
 */
export var LightModeBackgrounds;
(function (LightModeBackgrounds) {
    LightModeBackgrounds[LightModeBackgrounds["L1"] = 0] = "L1";
    LightModeBackgrounds[LightModeBackgrounds["L1Alt"] = 3] = "L1Alt";
    LightModeBackgrounds[LightModeBackgrounds["L2"] = 10] = "L2";
    LightModeBackgrounds[LightModeBackgrounds["L3"] = 13] = "L3";
    LightModeBackgrounds[LightModeBackgrounds["L4"] = 16] = "L4";
})(LightModeBackgrounds || (LightModeBackgrounds = {}));
/**
 * Friendly names for the indexes of dark mode backgrounds
 * on the neutral ramp. These values are designed to work with
 * the default neutral ramp provided by the DesignSystem
 *
 * @deprecated Use the recipes because they can be more dynamic for different ramps
 */
export var DarkModeBackgrounds;
(function (DarkModeBackgrounds) {
    DarkModeBackgrounds[DarkModeBackgrounds["L1"] = 76] = "L1";
    DarkModeBackgrounds[DarkModeBackgrounds["L1Alt"] = 76] = "L1Alt";
    DarkModeBackgrounds[DarkModeBackgrounds["L2"] = 79] = "L2";
    DarkModeBackgrounds[DarkModeBackgrounds["L3"] = 82] = "L3";
    DarkModeBackgrounds[DarkModeBackgrounds["L4"] = 85] = "L4";
})(DarkModeBackgrounds || (DarkModeBackgrounds = {}));
