// @microsoft/fast-components-react-base/dist/tabs/tab-panel.js
import Foundation from "@microsoft/fast-components-foundation-react";
import { classNames } from "@microsoft/fast-web-utilities";
import React from "react";
import { DisplayNamePrefix } from "../utilities";
class TabPanel extends Foundation {
    constructor() {
        super(...arguments);
        this.handledProps = {
            managedClasses: void 0,
            active: void 0,
            slot: void 0,
        };
    }
    /**
     * Renders the component
     */
    render() {
        return (React.createElement("div", Object.assign({}, this.unhandledProps(), { role: "tabpanel", "aria-hidden": !this.props.active, className: this.generateClassNames() }), this.props.children));
    }
    /**
     * Generates class names based on props
     */
    generateClassNames() {
        const { tabPanel, tabPanel__hidden, } = this.props.managedClasses;
        return super.generateClassNames(classNames(tabPanel, [tabPanel__hidden, !this.props.active]));
    }
}
TabPanel.defaultProps = {
    active: false,
    managedClasses: {},
};
TabPanel.displayName = `${DisplayNamePrefix}TabPanel`;
export default TabPanel;
