// @microsoft/fast-components-react-base/dist/tabs/tab.js
import Foundation from "@microsoft/fast-components-foundation-react";
import { classNames } from "@microsoft/fast-web-utilities";
import React from "react";
import { DisplayNamePrefix } from "../utilities";
class Tab extends Foundation {
    constructor() {
        super(...arguments);
        this.handledProps = {
            managedClasses: void 0,
            active: void 0,
            slot: void 0,
        };
    }
    /**
     * Renders the component
     */
    render() {
        return (React.createElement("div", Object.assign({}, this.unhandledProps(), { role: "tab", "aria-selected": this.props.active, className: this.generateClassNames() }), this.props.children));
    }
    /**
     * Generates class names based on props
     */
    generateClassNames() {
        const { tab, tab__active } = this.props.managedClasses;
        return super.generateClassNames(classNames(tab, [tab__active, this.props.active]));
    }
}
Tab.defaultProps = {
    active: false,
    managedClasses: {},
};
Tab.displayName = `${DisplayNamePrefix}Tab`;
export default Tab;
